pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// Reconcile reads that state of the cluster for a NSTemplateSet object and makes changes based on the state read
+// and what is in the NSTemplateSet.Spec
+func (r *ReconcileNSTemplateSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling NSTemplateSet")
+
+	// Fetch the NSTemplateSet instance
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, nsTmplSet)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	result, err := r.ensureNamespaces(reqLogger, nsTmplSet)
+	if err != nil || result.Requeue == true {
+		return result, err
+	}
+	return result, r.setStatusReady(nsTmplSet)
+}
+
+func (r *ReconcileNSTemplateSet) ensureNamespaces(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet) (reconcile.Result, error) {
+	userName := nsTmplSet.GetName()
+	labels := map[string]string{
+		"owner": userName,
+	}
+	opts := client.MatchingLabels(labels)
+	namespaces := &corev1.NamespaceList{}
+	if err := r.client.List(context.TODO(), opts, namespaces); err != nil {
+		return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err, "failed to list namespace with label owner '%s'", userName)
+	}
+
+	missingNs := findNsForProvision(namespaces.Items, nsTmplSet.Spec.Namespaces, userName)
+	if missingNs != (toolchainv1alpha1.Namespace{}) {
+		nsName := toNamespaceName(userName, missingNs.Type)
+		log.Info("provisioning namespace", "namespace", missingNs)
+		if err := r.setStatusNamespaceProvisioning(nsTmplSet, nsName); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		params := make(map[string]string)
+		params["USER_NAME"] = userName
+		err := r.applyTemplate(r.client, missingNs, params)
+		if err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s'", nsName)
+		}
+
+		// set labels
+		namespace := &corev1.Namespace{}
+		if err := r.client.Get(context.TODO(), types.NamespacedName{Name: nsName}, namespace); err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to get namespace '%s'", nsName)
+		}
+		if err := controllerutil.SetControllerReference(nsTmplSet, namespace, r.scheme); err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to set controller '%s'", nsName)
+		}
+		if namespace.Labels == nil {
+			namespace.Labels = make(map[string]string)
+		}
+		namespace.Labels["owner"] = userName
+		namespace.Labels["revision"] = missingNs.Revision
+		if err := r.client.Update(context.TODO(), namespace); err != nil {
```

Comment:
May I know why we cannot set the labels during the namespace provisioning - in case we use the creation of `namespace` instead of `ProjectRequest`?

Reply:
@MatousJobanek We will be using `ProjectRequest` only in template because when `Project` created with `ProjectRequest`, it will also create some extra roles.  Checked with @dipak-pawar on the same.

Reply:
@alexeykazakov label is manged by NSTemplateSetCtrl, it's the only actor which act upon it.  So maintaining in the scope of NSTemplateSetCtrl is better.  The other option is to set in `TemplateProcessor` cc: @xcoulon but to me TemplateProcessor should not bother about the such stuff which is used to calc whether to re-provision namespace or not.

Reply:
> it will also create some extra roles.

yes, it will add cluster-admin control over the resulting namespace that gets created (plus there can be an additional template for shaping the default skeleton of the namespace). But we don't need a special cluster-admin control as our SA already has it, and we don't need an additional template either (we will need it for namespaces created by advanced users, but that will be solved via this controller as well). The `ProjectRequest` is mainly focused on standard end-users, not cluster-admin service accounts that deal with specific use-cases. 

That said, I would rather try to use the direct creation of the namespaces with the labels already set. I think that it will be cleaner and it will be also simpler from the code point of view.

Reply:
Thinking about the default template that is applied when `ProjectRequest` is used. If we avoid it, then it could also simplify the logic that deals with conflicts in the process-apply part of the code @xcoulon